transformersjs/benchmark.js

```diff
@@ -0,0 +1,116 @@
+// Copyright 2025 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Polyfills that Transformers.js / the ONNX runtime needs in JavaScript shells.
+
+class URL {
+  href;
+  constructor(url, base) {
+    // DEBUG
+    // console.log('URL', url, base);
+    this.href = url;
+  }
+}
+globalThis.URL = URL;
+
+// Polyfill fetch for shell-compatibility and to cache / preload model weights etc.
+let preload = { /* Initialized in init() below due to async. */ };
+const originalFetch = globalThis.fetch ?? function(url) {
+  throw new Error("no fetch available");
+}
+globalThis.fetch = async function(url) {
+  // DEBUG
+  // console.log('fetch', url);
+
+  // Redirect some paths to cached/preloaded resources.
+  if (preload[url]) {
+    return {
+      ok: true,
+      status: 200,
+      arrayBuffer() { return preload[url]; },
+      async blob() {
+        return {
+          size: preload[url].byteLength,
+          async arrayBuffer() { return preload[url]; }
+        }
+      },
+    };
+  }
+
+  // This should only be called in the browser, where fetch() is available.
+  return originalFetch(url);
+};
+
+// JetStream benchmark harness. Reuse for two different Transformers.js tasks.
+// Assumes `preloadFiles(module)`, `initPipeline(pipelineFromTransformersJs)`,
+// and `doTask(initializedPipeline, inputArrayBuffer)` is in the global scope.
+
+class Benchmark {
+  transformersJsModule;
+  wasmBinary;
+  pipeline;
+  inputFile;
+  output;
+
+  async init() {
+    this.transformersJsModule = await JetStream.dynamicImport(JetStream.preload.transformersJsModule);
+    this.wasmBinary = await JetStream.getBinary(JetStream.preload.onnxWasmBinary);
+
+    for (const url of Object.values(JetStream.preload)) {
+      preload[url] = await JetStream.getBinary(url);
+    }
+
+    if ('inputFile' in JetStream.preload) {
+      this.inputFile = (await JetStream.getBinary(JetStream.preload.inputFile)).buffer;
+      // DEBUG
+      // console.log('inputFile', this.inputFile.byteLength, 'bytes');
+    }
+  }
+
+  async runIteration() {
+    // Initialize the inference pipeline in the first iteration.
+    if (!this.pipeline) {
+      // TODO: Profile startup only: What is taking so much time here?
+      let { env, pipeline } = this.transformersJsModule;
+    
+      env.allowRemoteModels = false;
+      env.allowLocalModels = true;
+      env.localModelPath = './transformersjs/build/models/';
+
+      // Always select the Wasm backend, nothing else.
+      delete env.backends.onnx.webgl;
+      delete env.backends.onnx.webgpu;
+    
+      // Single-threaded only for now, since we cannot spawn workers in shells.
+      // TODO: Implement sufficiently powerful workers in shells (or provide
+      // polyfills).
+      env.backends.onnx.wasm.numThreads = 1;
```

Comment:
How different is the workload if you enable threads? Were you ever able to get it to work? FWIW, I tried just changing this and I get an `Error: no available backend found.`